### PR TITLE
Added payload size check logic for Elasticsearch

### PR DESCRIPTION
**Pull request by sonika-shah**
### Describe your changes:

Added Payload size check logic in Elasticsearch

Fixes https://github.com/open-metadata/OpenMetadata/issues/15258

#
### Type of change:
- [x] Improvement

#
### Checklist:
- [x] I have read the [**CONTRIBUTING**](https://docs.open-metadata.org/developers/contribute) document.
- [x] My PR title is `Added payload size check logic for Elasticsearch`
- [x] I have commented on my code, particularly in hard-to-understand areas. 
- [x] For JSON Schema changes: I updated the migration scripts or explained why it is not needed.

### CODE REVIEW

File: openmetadata-service/src/main/java/org/openmetadata/service/search/elasticsearch/ElasticSearchIndexSink.java [hunk 1]
```diff
@@ -1,16 +1,17 @@
...
-import static org.openmetadata.service.workflows.searchIndex.ReindexingUtil.getUpdatedStats;
+import static org.openmetadata.service.workflows.searchIndex.ReindexingUtil.*;
 
+import es.org.elasticsearch.action.DocWriteRequest;
+import es.org.elasticsearch.action.bulk.BulkItemResponse;
 import es.org.elasticsearch.action.bulk.BulkRequest;
 import es.org.elasticsearch.action.bulk.BulkResponse;
 import es.org.elasticsearch.client.RequestOptions;
-import java.util.Map;
+import java.util.*;
```
Comment (harshach) [inline openmetadata-service/src/main/java/org/openmetadata/service/search/elasticsearch/ElasticSearchIndexSink.java:11]: avoid wild-card imports

File: openmetadata-service/src/main/java/org/openmetadata/service/apps/bundles/searchIndex/SearchIndexApp.java [hunk 2]
```diff
@@ -252,6 +261,12 @@ private void dataInsightReindex(JobExecutionContext jobExecutionContext) {
       while (!stopped && !paginatedDataInsightSource.isDone()) {
         try {
           resultList = paginatedDataInsightSource.readNext(null);
+          List<String> entityName =
+              resultList.getData().stream()
+                  .map(entity -> paginatedDataInsightSource.getEntityType() + " " + entity.getId())
+                  .collect(Collectors.toList());
```
Comment (mohityadav766) [inline openmetadata-service/src/main/java/org/openmetadata/service/apps/bundles/searchIndex/SearchIndexApp.java:267]: You can use string.format here

File: openmetadata-service/src/main/java/org/openmetadata/service/search/elasticsearch/ElasticSearchIndexSink.java [hunk 3]
```diff
@@ -33,29 +36,80 @@ public BulkResponse write(BulkRequest data, Map<String, Object> contextData)
       throws SearchIndexException {
     LOG.debug("[EsSearchIndexSink] Processing a Batch of Size: {}", data.numberOfActions());
     try {
-      BulkResponse response = searchRepository.getSearchClient().bulk(data, RequestOptions.DEFAULT);
-      int currentSuccess = getSuccessFromBulkResponseEs(response);
-      int currentFailed = response.getItems().length - currentSuccess;
+      BulkResponse response = null;
+      int offset = 0;
```
Comment (harshach) [inline openmetadata-service/src/main/java/org/openmetadata/service/search/elasticsearch/ElasticSearchIndexSink.java:40]: what is offset trackign? is it the no.of entities we are going to index?

Comment (sonika-shah) [reply openmetadata-service/src/main/java/org/openmetadata/service/search/elasticsearch/ElasticSearchIndexSink.java:40]: purpose of offset tracking is to maintain the bufferData until it has not reached the payload limit inside the batch.

Comment (mohityadav766) [reply openmetadata-service/src/main/java/org/openmetadata/service/search/elasticsearch/ElasticSearchIndexSink.java:45]: @harshach number of records reaching sink are batched on the basis of their size , offset is used to track the records in the main list that sink receives

File: conf/openmetadata.yaml [hunk 4]
```diff
@@ -256,6 +256,7 @@ elasticsearch:
   socketTimeoutSecs: ${ELASTICSEARCH_SOCKET_TIMEOUT_SECS:-60}
   keepAliveTimeoutSecs: ${ELASTICSEARCH_KEEP_ALIVE_TIMEOUT_SECS:-600}
   batchSize: ${ELASTICSEARCH_BATCH_SIZE:-10}
+  payLoadSize: ${ELASTICSEARCH_PAYLOAD_BYTES_SIZE:-10485760}   #max payLoadSize in Bytes
```
Comment (mohityadav766) [inline conf/openmetadata.yaml:259]: @sonika-shah we should add this ENV to docker files as well

File: openmetadata-service/src/main/java/org/openmetadata/service/search/elasticsearch/ElasticSearchIndexSink.java (hunk 3 above)
Comment (mohityadav766) [reply openmetadata-service/src/main/java/org/openmetadata/service/search/elasticsearch/ElasticSearchIndexSink.java:45]: batchSize -> Number of Records to read from Db in one go
payloadSize -> max size of payload that can be sent in one go to ES, max number of entities that can be sent = batchSize , or else whenever we hit the payload size it's flushed

File: openmetadata-service/src/main/java/org/openmetadata/service/apps/bundles/searchIndex/SearchIndexApp.java [hunk 5]
```diff
@@ -217,6 +216,16 @@ private void entitiesReIndex(JobExecutionContext jobExecutionContext) {
       while (!stopped && !paginatedEntitiesSource.isDone()) {
         try {
           resultList = paginatedEntitiesSource.readNext(null);
+          List<String> entityName =
+              resultList.getData().stream()
```
[the remark this answers is not in the document]
Comment (sonika-shah) [reply openmetadata-service/src/main/java/org/openmetadata/service/apps/bundles/searchIndex/SearchIndexApp.java:223]: Thank you for your suggestion. While I appreciate the concern, I believe creating a new function for a simple stream call may add unnecessary complexity. Let's keep it inline for clarity.

File: openmetadata-service/src/main/java/org/openmetadata/service/search/elasticsearch/ElasticSearchIndexSink.java [hunk 6]
```diff
@@ -33,29 +41,80 @@ public BulkResponse write(BulkRequest data, Map<String, Object> contextData)
...
-        throw new SearchIndexException(
-            new IndexingError()
-                .withErrorSource(SINK)
-                .withSubmittedCount(data.numberOfActions())
-                .withSuccessCount(currentSuccess)
-                .withFailedCount(currentFailed)
-                .withMessage("Issues in Sink To Elastic Search.")
-                .withFailedEntities(getErrorsFromBulkResponse(response)));
-      }
+      while (offset != data.requests().size()) { // until all requests in the batch are processed
```
Comment (harshach) [inline openmetadata-service/src/main/java/org/openmetadata/service/search/elasticsearch/ElasticSearchIndexSink.java:54]: what are these two loops for?

**Comment by harshach** [comment]
@mohityadav766 @sonika-shah is the feedback incorporated into the PR?

**Review by mohityadav766 — APPROVED** [review]

**Pull request merged**